[Bug] AAP 42572 database deadlock

djyasin opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This PR resolves AAP-422572 Database deadlock by awx_callback_receiver_worker and awx_dispatcher_worker while trying to update hosts last_job_id and ansible_facts in two separate commands
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 24.6
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

AlanCoding committed:
Demo of sorting hosts

AlanCoding committed:
Comment out actual fix

djyasin committed:
Sort both bulk updates and add batch size to facts bulk update to resolve deadlock issue

djyasin committed:
Remove del load_credentials to resolve CI issue

djyasin committed:
Update tests to expect batch_size to agree with changes

djyasin commented:
@AlanCoding I've added sorting before both bulk updates and have added  a matching batch size to the facts.py bulk update.

It seems like regardless of the sorting, the deadlock occurs if the batch size is present.

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -62,7 +62,8 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
 def raw_update_hosts(host_list):
-    Host.objects.bulk_update(host_list, ['ansible_facts', 'ansible_facts_modified'])
+    host_list = sorted(host_list, key=lambda host: host.id)
+    Host.objects.bulk_update(host_list, ['ansible_facts', 'ansible_facts_modified'], batch_size=100)
```
If the fix here is legitimate (otherwise why do it?) then we should get rid of the `except OperationalError as exc:` later down here.

AlanCoding commented on awx/main/models/events.py:
```diff
@@ -602,7 +602,7 @@ def _update_host_summary_from_stats(self, hostnames):
                     updated_hosts.add(h)
 
-            Host.objects.bulk_update(list(updated_hosts), ['last_job_id', 'last_job_host_summary_id'], batch_size=100)
+            Host.objects.bulk_update(sorted(updated_hosts, key=lambda host: host.id), ['last_job_id', 'last_job_host_summary_id'], batch_size=100)
```
For longer-term maintainability, it might be nice to introduce a utility method. Even if the method seems trivial (a simple `sorted` action), it would help stitch these two pieces of code together, and establish the expectation for the future.

And actually, I might even use that here:

https://github.com/ansible/awx/blob/db6e8b9bad660bdc028f48e8461e6794b6c3cef1/awx/main/tasks/host_metrics.py#L150

That might not _currently_ risk contention with other queries, but it is a bulk_update with a batch_size. Seems like there may be a good argument that _all_ such queries in the app should use consistent ordering, because it completely fixes the Deadlock issue.

That would also allow re-using that method in the test here.

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -62,7 +62,8 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
+    Host.objects.bulk_update(host_list, ['ansible_facts', 'ansible_facts_modified'], batch_size=100)
 
 
 def update_hosts(host_list, max_tries=5):
```
Later in this file, you can see a loop:

```python
for host in hosts_cached:
```

The calls to `update_hosts` are _manually batched_, and I would expect that causes the same issue as Django's `batch_size`. This is using a batch size inside of a batch. You might as well get rid of the `batch_size` kwarg.

As you do the main loop over `hosts_cached`, I expect the hosts need to be sorted there.

djyasin committed:
Add utility method to bulk update and sort hosts and applied that to the appropriate locations

Remove unused imports

Add utility method for sorting bulk updates

Remove try except OperationalError for loop

Remove unused import of django.db.OperationalError

Remove batch size as it is now on the bulk update utility method as 100

Remove batch size here since it is specified in sortedbulkupdate

Add transaction.atomic to have entire transaction is run as a signle transaction before committing to the db

Revert change to bulk update as it's not needed here and just sort instead

Move bulk_sorted utility method into db.py and updated name to not be specific to Hosts

Revise to import bulk_update_sorted.. rather than calling it as an argument

Fix way I'm importing bulk_update_sorted.. Remove unneeded Host import and remove calls to bul_update as args

Rebise calls to bulk_update_sorted.. to include Host in the args

REmove raw_update_hosts method and replace with bulk_update_sorted_by_id in update_hosts

Remove update_hosts function and replace with  bulk_update_sorted_by_id

Update live tests to use bulk_update_sorted_by_id

Fix the fields in bulk_update to agree with test

AlanCoding commented on awx/main/sortedbulkupdate.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2015 Ansible, Inc.
```
I'd like this in a different folder. See `awx/main/utils/mem_inventory.py` as an example, a new file like `awx/main/utils/inventory.py` might be okay, or even adding it to `awx/main/utils/db.py` actually would work fine in my book (because it is a DB utility).

My original thought is that this doesn't have to be specific to the `Host` model. Yeah, you import the `Host` model, but you don't have to. The class can be taken as an argument or obtained indirectly from the type of the objects.

AlanCoding commented on awx/main/models/events.py:
```diff
@@ -536,7 +536,7 @@ def _hostnames(self):
         return hostnames
 
-    def _update_host_summary_from_stats(self, hostnames):
+    def _update_host_summary_from_stats(self, hostnames, bulk_update_hosts_sorted_by_id):
```
It looks like `bulk_update_hosts_sorted_by_id` is your utility method. It would make more sense to import that method as opposed to taking it as an argument.

Guessing, but you might have had problems with circular imports. The thought process should be to keep a mental model or "order" of models. Utility functions should have the fewest (ideally no) internal dependencies, so I would point to my other comment that the module with the utility method shouldn't do `from awx.main.models.inventory import Host` and this should address any circularity you had importing that method from _this_ module.

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -61,26 +61,14 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
-            else:
-                raise
-        break
+    raw_update_hosts(Host, host_list, bulk_update_sorted_by_id)
```
Best to delete the method `raw_update_hosts` and put here

```python
bulk_update_sorted_by_id(host_list, fields=['ansible_facts', 'ansible_facts_modified'])
```

AlanCoding commented on awx/main/tasks/facts.py:
```diff
@@ -137,4 +127,4 @@ def finish_fact_cache(hosts_cached, destination, facts_write_time, log_data, job
             update_hosts(hosts_to_update)
             hosts_to_update = []
-    update_hosts(hosts_to_update)
+    update_hosts(hosts_to_update, bulk_update_sorted_by_id)
```
Actually, I would say also delete `update_hosts` and put that line here.

AlanCoding commented on awx/main/tasks/host_metrics.py:
```diff
@@ -119,7 +119,7 @@ def __init__(self):
         self.records_to_update = []
 
-    def execute(self):
+    def execute(self, bulk_update_sorted_by_id):
```
Here too, I think you should not add this as an argument. If it can't be imported, then address that by modifying the `bulk_update_sorted_by_id` method / module.

AlanCoding commented on awx/main/utils/db.py:
```diff
@@ -4,7 +4,22 @@
     set_application_name(settings.DATABASES, settings.CLUSTER_HOST_ID, function=function)
+
+
+def bulk_update_sorted_by_id(model, objects, fields, batch_size=1000):
```
This appears inconsistent with your use of the method `bulk_update_sorted_by_id(host_list, fields=['ansible_facts', 'ansible_facts_modified'])`.

So the next implied question would be "how should this get `Host`?"

You can pass in `Host`, the class, that would be fine if you do that (use the current method arg spec). Just update the method calls if you do.

The other alternative would be that after you define `sorted_objects`, you set `model = type(sorted_objects[0])`. Some might find this hacky, but I'm totally okay with it. I do worse import workarounds myself. All the time.

AlanCoding replied:
It would be really good to add a docstring basically explaining the bug that you're fixing and how this addresses it. That establishes the reasons for the method's existence.

AlanCoding commented on awx/main/tests/live/tests/test_host_update_contention.py:
```diff
@@ -0,0 +1,79 @@
+    # host_list = sorted(host_list, key=lambda host: host.id)
+
+    # NOTE: did not reproduce the bug without batch_size
+    Host.objects.bulk_update(host_list, [field_name], batch_size=100)
```
By now I think it's well-established that this test is failing. So import & use your new utility method, and I'd like to see confirmation of the fix in the github checks.

AlanCoding commented:
Documenting the test fail here, so we can confirm fix later:

```
_________________________ test_host_update_contention __________________________

default_org = <Organization: Default-1>

    def test_host_update_contention(default_org):
        inv_kwargs = dict(organization=default_org, name='test_host_update_contention')
    
        if Inventory.objects.filter(**inv_kwargs).exists():
            inv = Inventory.objects.get(**inv_kwargs).delete()
    
        inv = Inventory.objects.create(**inv_kwargs)
        right_now = now()
        hosts = [Host(inventory=inv, name=f'host-{i}', created=right_now, modified=right_now) for i in range(1000)]
        print('bulk creating hosts')
        Host.objects.bulk_create(hosts)
    
        # sanity check
        for host in hosts:
            assert not host.variables
    
        # Force our worker pool to make their own connection
        connection.close()
    
        ready_events = [multiprocessing.Event() for _ in range(2)]
        continue_event = multiprocessing.Event()
    
        print('spawning processes for concurrent bulk updates')
        processes = []
        fields = ['variables', 'ansible_facts']
        for i in range(2):
            p = multiprocessing.Process(target=worker_delete_target, args=(ready_events[i], continue_event, fields[i]))
            processes.append(p)
            p.start()
    
        # Assure both processes are connected and have loaded their host list
        for e in ready_events:
            print('waiting on subprocess ready event')
            e.wait()
    
        # Begin the bulk_update queries
        print('setting the continue event for the workers')
        continue_event.set()
    
        # if a Deadloack happens it will probably be surfaced by result here
        print('waiting on the workers to finish the bulk_update')
        for p in processes:
            p.join()
    
        print('checking workers have variables set')
        for host in inv.hosts.all():
>           assert host.variables.startswith('my_var:')
E           AssertionError: assert False
E            +  where False = <built-in method startswith of str object at 0x7f44ae002a60>('my_var:')
E            +    where <built-in method startswith of str object at 0x7f44ae002a60> = ''.startswith
E            +      where '' = <Host: host-0-12>.variables

tests/test_host_update_contention.py:78: AssertionError
----------------------------- Captured stdout call -----------------------------
bulk creating hosts
spawning processes for concurrent bulk updates
waiting on subprocess ready event
worker has loaded all the hosts needed
waiting on subprocess ready event
worker has loaded all the hosts needed
setting the continue event for the workers
waiting on the workers to finish the bulk_update
finished doing the bulk update in worker
checking workers have variables set
----------------------------- Captured stderr call -----------------------------
Process Process-1:
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/psycopg/cursor.py", line 97, in execute
    raise ex.with_traceback(None)
psycopg.errors.DeadlockDetected: deadlock detected
DETAIL:  Process 358 waits for ShareLock on transaction 2007; blocked by process 359.
Process 359 waits for ShareLock on transaction 2008; blocked by process 358.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (1,10) in relation "main_host"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib64/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib64/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/awx_devel/awx/main/tests/live/tests/test_host_update_contention.py", line 28, in worker_delete_target
    Host.objects.bulk_update(host_list, [field_name], batch_size=100)
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/models/query.py", line 892, in bulk_update
    rows_updated += queryset.filter(pk__in=pks).update(**update_kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/models/query.py", line [120](https://github.com/ansible/awx/actions/runs/14645470181/job/41098399679#step:5:121)6, in update
    rows = query.get_compiler(self.db).execute_sql(CURSOR)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/models/sql/compiler.py", line 1984, in execute_sql
    cursor = super().execute_sql(result_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/models/sql/compiler.py", line 1562, in execute_sql
    cursor.execute(sql, params)
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/backends/utils.py", line 102, in execute
    return super().execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/backends/utils.py", line 84, in _execute
    with self.db.wrap_database_errors:
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/psycopg/cursor.py", line 97, in execute
    raise ex.with_traceback(None)
django.db.utils.OperationalError: deadlock detected
DETAIL:  Process 358 waits for ShareLock on transaction 2007; blocked by process 359.
Process 359 waits for ShareLock on transaction 2008; blocked by process 358.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (1,10) in relation "main_host"
```

djyasin committed:
Update functional tests to use bulk_update_sorted_by_id since update_hosts has been deleted

Replace update_hosts with bulk_update_sorted_by_id

Remove referenes to update_hosts

Update corresponding fact cachin tests to use bulk_update_sorted_by_id

Remove import of bulk_sorted_update

Add code comment to live test to silence Sonarqube hotspot

djyasin committed:
Add comment NOSONAR to get rid of Sonarqube warning since this is just a test and it's not actually a security issue

Get test_finish_job_fact_cache_with_existing_data passing

Get test_finish_job_fact_cache_clear passing

Remove reference to raw_update and replace with new bulk update utility method

Add pytest.mark.django_db to appropriate tests

Corrent which model is called in bulk_update_sorted_by_id

Remove now unused Host import

Point to where bulk_update_sorted_by_id to where that is actually being used

Correct import of bulk_update_sorted_by_id

Revert changes in this file to avoid db calls issue

Remove @pytest.mark.django_db from unit tests

Remove commented out host sorting suggested fix

Fix failing tests test_pre_post_run_hook_facts_deleted_sliced & test_pre_post_run_hook_facts

Remove atomic transaction line, add return, and add docstring

AlanCoding commented on awx/main/tests/unit/models/test_jobs.py:
```diff
@@ -58,6 +63,7 @@ def test_fact_cache_with_invalid_path_traversal(tmpdir):
     assert os.listdir(os.path.join(fact_cache, '..')) == ['facts']
 
 
+@pytest.mark.django_db
```
We shouldn't use this for any tests under `awx/main/tests/unit`. That's the idea of the separation of folders. If the test doesn't need the database, don't add the marker. If it does, then move to `awx/main/tests/functional`.

AlanCoding commented on awx/main/tests/live/tests/test_host_update_contention.py:
```diff
@@ -0,0 +1,81 @@
+    continue_event.wait()
+
+    # # presumed fix
+    # host_list = sorted(host_list, key=lambda host: host.id)
```
You should be able to remove this comment now, since the test now uses `bulk_update_sorted_by_id` and the dev-env check is passing. The test output shows it ran the 1 test in this file.

AlanCoding commented on awx/main/utils/db.py:
```diff
@@ -4,7 +4,22 @@
+    sorted_objects = sorted(objects, key=lambda obj: obj.id)
+
+    # Perform the bulk update within an atomic transaction
+    with transaction.atomic():
```
The bulk update is a single statement, a transaction should not be necessary at all. The task system should usually, as a best practice, operate in auto-commit mode. If you don't, the read-write flow behaves in a way that's not intuitive and makes cross-process errors more likely due to uncommited changes in another client's transaction.

So I think the best course of action is to just not do this in a transaction.

AlanCoding commented on awx/main/utils/db.py:
```diff
@@ -4,7 +4,22 @@
+
+    # Perform the bulk update within an atomic transaction
+    with transaction.atomic():
+        model.objects.bulk_update(sorted_objects, fields, batch_size=batch_size)
```
https://docs.djangoproject.com/en/5.2/ref/models/querysets/#bulk-update

bulk_update has a return value. I don't care what the value is, but you should `return` it.

djyasin replied:
How does this look to you now?

AlanCoding reviewed (approved):
Current code state seems to be fully matured, with plenty sufficient associated tests within the patch.

djyasin committed:
Fix failing test test_finish_job_fact_cache_clear & test_finish_job_fact_cache_with_existing_data
